base/sparse/linalg.jl

```diff
@@ -102,6 +102,34 @@ function (*){TX,TvA,TiA}(X::StridedMatrix{TX}, A::SparseMatrixCSC{TvA,TiA})
     Y
 end
 
+function (*){TvA,TiA}(X::Diagonal, A::SparseMatrixCSC{TvA,TiA})
+    mX, nX = size(X)
+    if nX != A.m
+        throw(DimensionMismatch("second dimension of X, $nX, must match first dimension of A, $(A.m)"))
+    end
+    Ynzval = zeros(promote_type(eltype(X),TvA), length(A.nzval))
+    rowval = A.rowval
+    nzval = A.nzval
+    @inbounds for k=1:length(nzval)
+        Ynzval[k] = X[rowval[k],rowval[k]] * nzval[k]
```

Comment:
Probably faster to access the underlying vector directly rather than doing `X[i,i]`